### DISCUSSION TITLE
Comments: Learn Enhanced C#

### DISCUSSION THREAD

**Discussion by qwertie** [qwertie/ecsharp#28] [Web site comments]
Comments on [Enhanced C# for ordinary coders](http://ecsharp.net/ecs/for-normal-coders.html)

**Comment by gjroelofs** [comment]
Thank you for your work @qwertie, and I am actually deciding whether to integrate this library into our project.
The two primary applications we're looking at are ADTs for Address Types on spatial datastructures and value types for the different types of coordinate systems. Would you have any tips on what might help us on the latter?  

(I.e.: presumably wrapped float & int Vector3 with the option to preserve (or forward to) all API that exists on the wrapped types, but with type checking on conversion.)

**Comment by qwertie** [comment]
Well, I should point out that the Loyc.Math library includes `Vector3<T>` and `Point3<T>` types, but that doesn't sound like what you're looking for.

As LeMP is just a preprocessor, it doesn't do semantic analysis, so it doesn't know anything about code located elsewhere, i.e. the wrapped type, so it doesn't include a feature to "forward to all APIs in a wrapped type". The `==>` operator is helpful for setting up method forwarding quickly, as the article explains, but currently it doesn't support explicit type conversions. You could use implicit conversions, as in

    class WrappedType { TypeX Value; }

    class WrapperType {
        public WrapperType(private WrappedType _wrapper) {}
        TypeThatConvertsImplicitlyFromTypeX Value ==> _wrapper.#;
    }

If that's not sufficient, let me know and I'll give it some thought.
